cdap-ui/app/cdap/components/FastAction/SetPreferenceAction/index.js

```diff
@@ -0,0 +1,323 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import React, {Component, PropTypes} from 'react';
+import FastActionButton from '../FastActionButton';
+import isObject from 'lodash/isObject';
+import upperFirst from 'lodash/upperFirst';
+import T from 'i18n-react';
+import shortid from 'shortid';
+import { Modal, Tooltip, ModalHeader, ModalBody } from 'reactstrap';
+import myPreferenceApi from 'api/preference';
+import KeyValuePairs from 'components/KeyValuePairs';
+import NamespaceStore from 'services/NamespaceStore';
+require('./SetPreferenceAction.less');
+
+export default class SetPreferenceAction extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      modal: false,
+      loading: false,
+      errorMessage: '',
+      extendedMessage: '',
+      error: '',
+      preferences: [{}],
+      keyValues: {'pairs': [{'key':'', 'value':'', 'uniqueId': ''}]},
+      inheritedPreferences: []
+    };
+
+    this.eventText = '';
+    this.appId = this.props.entity.id;
+    this.toggleModal = this.toggleModal.bind(this);
+    this.onKeyValueChange = this.onKeyValueChange.bind(this);
+    this.toggleTooltip = this.toggleTooltip.bind(this);
+
+  }
+
+  toggleTooltip() {
+    this.setState({ tooltipOpen : !this.state.tooltipOpen });
+  }
+  toggleModal(event) {
+    this.setState({
+      modal: !this.state.modal,
+      loading: false,
+      errorMessage: '',
+      extendedMessage: '',
+      textarea: false,
+      textInput: '',
+      droppedFile: {}
+    });
+    if (event) {
+      event.stopPropagation();
+      event.nativeEvent.stopImmediatePropagation();
+    }
+  }
+
+  getAppPreferences() {
+    let {selectedNamespace: namespace} = NamespaceStore.getState();
+    myPreferenceApi.getAppPreference({namespace, appId: this.appId})
+    .subscribe(
+      (res) => {
+        if (!Object.keys(res).length) {
+          this.setState({keyValues: {'pairs': [{'key':'', 'value':'', 'uniqueId': shortid.generate()}]}});
+        } else {
+          this.setState({keyValues: {'pairs': this.getKeyValPair(res)}});
+        }
+      },
+      (error) => {
+        this.setState({
+          error: isObject(error) ? error.response : error
+        });
+      }
+    );
+  }
+
+  getInheritedPreferences() {
+    let {selectedNamespace: namespace} = NamespaceStore.getState();
+    myPreferenceApi.getSystemPreference()
+    .subscribe(
+     (systemRes) => {
+       let systemPrefArray = this.getKeyValPair(systemRes, 'system');
+       myPreferenceApi.getNamespacePreference({namespace})
+      .subscribe(
```

Comment:
I don't think we need to chain at all. We should be able to get the resolved preference by using 1 API with `resolved=true` query parameter for namespace level.